core/frontend.ml

```diff
@@ -25,89 +25,178 @@ struct
-        ->- DesugarDatatypes.sentence tyenv
-        ->- uncurry TypeSugar.Check.sentence
-        (*  ->- after_typing ((FixTypeAbstractions.fix_type_abstractions tyenv)#sentence ->- snd)*)
-        ->- after_typing ((DesugarCP.desugar_cp tyenv)#sentence ->- snd)
-        ->- after_typing ((DesugarInners.desugar_inners tyenv)#sentence ->- snd)
-        ->- after_typing ((DesugarProcesses.desugar_processes tyenv)#sentence ->- snd)
-        ->- after_typing ((DesugarFors.desugar_fors tyenv)#sentence ->- snd)
-        ->- after_typing ((DesugarRegexes.desugar_regexes tyenv)#sentence ->- snd)
-        ->- after_typing ((DesugarFormlets.desugar_formlets tyenv)#sentence ->- snd)
-        ->- after_typing ((DesugarPages.desugar_pages tyenv)#sentence ->- snd)
-        ->- after_typing ((DesugarFuns.desugar_funs tyenv)#sentence ->- snd))
-    in
-    apply sentence
-
-let interactive tyenv pos_context sentence =
+    Debug.print ("Post-Frontend AST:\n" ^ Sugartypes.show_program result_program);
+
+  (result_program, t, cur_tyenv), ffi_files
+
+
+  type pre_typing_sentence_transformer =
+    Sugartypes.sentence -> Sugartypes.sentence
+
+  type post_typing_sentence_transformer =
+    Types.typing_environment -> Sugartypes.sentence -> Sugartypes.sentence
+
+  (* Program transformations before type-checking on sentences (i.e., REPL mode *)
+  let interactive_pre_typing_transformers
+    (pos_context : SourceCode.source_code)
+    (prev_tyenv : Types.typing_environment) (* type env before checking current program *)
+      : pre_typing_sentence_transformer list =
+    [
+      (ResolvePositions.resolve_positions pos_context)#sentence;
+      for_side_effects
+        CheckXmlQuasiquotes.checker#sentence;
+      DesugarModules.desugar_sentence;
+      DesugarLAttributes.desugar_lattributes#sentence;
+      LiftRecursive.lift_funs#sentence;
+      DesugarDatatypes.sentence prev_tyenv;
+    ]
+
+
+  (* Program transformations after type-checking sentences (i.e., REPL mode *)
+  let interactive_post_typing_transformers
+      : post_typing_sentence_transformer list =
+    [
+      DesugarCP.desugar_sentence;
+      DesugarInners.desugar_sentence;
+      DesugarProcesses.desugar_sentence;
+      DesugarFors.desugar_sentence;
+      DesugarRegexes.desugar_sentence;
+      DesugarFormlets.desugar_sentence;
+      DesugarPages.desugar_sentence;
+      DesugarFuns.desugar_sentence;
+    ]
+
+
+let interactive prev_tyenv pos_context sentence =
   if Settings.get_value Basicsettings.show_pre_frontend_ast then
     Debug.print ("Pre-Frontend AST:\n" ^ Sugartypes.show_sentence sentence);
 
-  let (post_sentence, _, _) as result = interactive_pipeline tyenv pos_context sentence in
+  let pre_transformers =
+    interactive_pre_typing_transformers pos_context prev_tyenv in
+  let apply_pre_tc_transformer sentence transformer =
+    transformer sentence in
+  let pre_tc_sentence =
+    List.fold_left
+      apply_pre_tc_transformer
+      sentence
+      pre_transformers in
```

Comment:
What's up with these trailing `in`s? And why is a linear piece separated by spaces? At first glance, this looks like the code is not connected. I find that highly confusing to read.